### PR TITLE
Doc update to introduce Tiered Platform Support Model

### PR DESCRIPTION
**Pull request by rrschulze**
**Description:** This PR adds documentation to introduce a tiered platform support model for the OpenTelemetry Collector.  The tiered platform support model provides clarity to the project and its users about how existing platforms are supported today and how requests for new platforms can be supported in future, while balancing between the aim to support as many platforms as possible and to guarantee stability for the most important platforms.

**Link to tracking Issue:** #8209

### CODE REVIEW

File: docs/platform-support.md
Comment (mx-psi) [inline docs/platform-support.md:19]: I think our support for Kubernetes comes from specific components that provide functionality related to it, but there is nothing special on the core code to support this. In contrast, to support an operating system family there are changes as to what system APIs are used. So I would be leaning towards not including Kubernetes within this document, or at least separating it somehow.

File: docs/platform-support.md
Comment (mx-psi) [inline docs/platform-support.md:1]: IMO we should define what the different platforms are. To that end I think we can just say they are a pair consisting of an operating system family together with a processor architecture family as defined by the Go team through the GOOS and GOARCH tags. We can also comment on this area as to what operating system versions we support (specifically, I would say as a library we support all OS versions officially supported by Go on their supported versions, and as a binary all the ones on the latest version of Go).

cc @rmfitzpatrick on this, please comment if you think the OS versions supported should be different.

File: docs/platform-support.md
Comment (mx-psi) [inline docs/platform-support.md:33]: I would use `darwin` here and generally stick to the naming given by the Go team.

File: docs/platform-support.md
Comment (mx-psi) [inline docs/platform-support.md:18]: I think we should state an owner for each platform supported. For Tier 1 platforms this can be the OTel collector approvers (at least for now), and we can leave the rest empty and create an issue to add owners based on who contributed PRs related to it.

File: docs/platform-support.md
Comment (mx-psi) [inline docs/platform-support.md:15]: We should state whether a defect specific to a platform of each tier can block the release process or warrant a bugfix release (I think for tier 1 they can).

Comment (mx-psi) [inline docs/platform-support.md:15]: I think this would benefit from a table form presentation, where we have columns for each aspect (CI testing, running tests, bugfix releases, binaries availability, possibility of dummy implementations....) and rows for each tier.

File: docs/platform-support.md
Comment (mx-psi) [inline docs/platform-support.md:6]: I would make this less concrete as to what specific versions we use for testing, outlining instead how we decide these versions. For CI testing this has de facto been "the latest OS version supported on Github runners".

File: docs/platform-support.md
Comment (mx-psi) [inline docs/platform-support.md:15]: Another requirement for tier 1 platforms in my mind would be the need to have a non-dummy implementation for any feature that is not platform specific. That is, we will not accept PRs that do not implement functionality on e.g. Linux/amd64.

File: docs/platform-support.md
Comment (mx-psi) [inline docs/platform-support.md:39]: I would expect Windows/amd64 to be at least Tier 2, why do you think it is tier 1?

File: docs/platform-support.md
Comment (mx-psi) [inline docs/platform-support.md:9]: We should say what guarantees we give to platforms that we give no support to officially. I think https://github.com/open-telemetry/opentelemetry-collector/pull/6924#pullrequestreview-1250396643 could be a good start.

File: docs/platform-support.md
Comment (mx-psi) [inline docs/platform-support.md:7]: I would leave packaging and installation methods out of this part. They are not strictly related to testing and we recently added wording here https://github.com/open-telemetry/opentelemetry-collector-releases/tree/main/distributions#criteria-for-supported-distributions as to what artifacts we promise to release.

File: docs/platform-support.md
Comment (mx-psi) [inline docs/platform-support.md:1]: Maybe not on this document/PR, but would also be good to define the process by which a platform can move between tiers (probably needs Collector approver/maintainer sponsorship?)

**Review by mx-psi — COMMENTED** [review]
Thanks for writing this up! I did a first review; will be out for a couple weeks and take this back when I return :)

File: docs/platform-support.md
Comment (aunshc) [inline docs/platform-support.md:44]:
```suggestion
The proposed additional platforms Linux on s390x ([#378](https://github.com/open-telemetry/opentelemetry-collector-releases/issues/378#issuecomment-1662371014)) and AIX on ppc64 ([#19195](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19195#issuecomment-1458560971)) will be included into Tier 3 once they're added to the OpenTelemetry Collector as platforms. 
```

File: docs/platform-support.md
Comment (aunshc) [inline docs/platform-support.md:23]:
```suggestion
Tier 2 platforms are _guaranteed to work with specified limitations_. Precompiled binaries are built and tested on the platform as part of the biweekly release cycle. Build and test infrastructure is provided by the platform maintainers. All tests are executed on the platform as far as they are applicable, and all prerequisites are fulfilled. Not executed tests and not tested collector add-ons (receivers, processors, exporters, etc.) are published on release of the collector distribution. Any build or test failure delays the release of the binaries for the respective platform but not the collector distribution for all other platforms. Defects are addressed but not with the priority as for Tier 1 and, if specific to the platform, require the support of the platform maintainers.
```

File: docs/platform-support.md
Comment (rrschulze) [reply docs/platform-support.md:44]: Instead of linking to specific issue comments here, it is sufficient to link to the related issues. I will update the link for AIX on ppc64.

File: docs/platform-support.md
Comment (rrschulze) [reply docs/platform-support.md:1]: Added definition of platform in this section. With respect to a generic support statement for OS and binaries need to discuss further, as it is not clear to me yet how this relates to test strategy and association with support tiers.

Comment (rrschulze) [reply docs/platform-support.md:1]: Can we assume that the assignment of a platform to a specific support tier is fully and only dependent on how the platform is tested as per test strategy? Can we further assume that a platform is elevated to an higher support tier once all required tests for this tier are coded, automatically executed and successfully completed? In this cases, we may tie the description for moving to an higher tier together with the test strategy. In-fact, maybe reasonable to separate out the test strategy into another document.

Comment (mx-psi) [reply docs/platform-support.md:1]: > it is not clear to me yet how this relates to test strategy and association with support tiers

My point is just that if we say we support Linux/Windows, it's unclear what versions of that OS family we support and we should say that as part of the definition of each platform.

Comment (mx-psi) [reply docs/platform-support.md:1]: I think there are concerns that are not strictly related to testing, e.g. the ones I mention on https://github.com/open-telemetry/opentelemetry-collector/pull/8224#discussion_r1293645840 and https://github.com/open-telemetry/opentelemetry-collector/pull/8224#discussion_r1293642032. Maybe we can tie these to testing, but I am unsure about it (e.g. I don't think with our current release schedule a bug on a well-tested but very niche platform would warrant a bugfix release).

File: docs/platform-support.md
Comment (rrschulze) [reply docs/platform-support.md]: Removed references to Kubernetes and Docker.

File: docs/platform-support.md
Comment (rrschulze) [reply docs/platform-support.md:1]: Added statement about supported OS versions based on Go support.

File: docs/platform-support.md
Comment (rrschulze) [reply docs/platform-support.md]: Updated platform naming.

File: docs/platform-support.md
Comment (rrschulze) [reply docs/platform-support.md]: Added table with owners column to supported platforms.

File: docs/platform-support.md
Comment (rrschulze) [reply docs/platform-support.md]: Added summary table.

File: docs/platform-support.md
Comment (rrschulze) [reply docs/platform-support.md:6]: Removed references to specific OS versions.

File: docs/platform-support.md
Comment (rrschulze) [reply docs/platform-support.md]: Moved windows/amd64 to Tier 2 and clarified Tier 2 requirements.

File: docs/platform-support.md
Comment (rrschulze) [reply docs/platform-support.md:18]: Added statement on unsupported platforms.

File: docs/platform-support.md
Comment (rrschulze) [reply docs/platform-support.md]: Clean up done of the test strategy section.

File: docs/platform-support.md
Comment (mx-psi) [inline docs/platform-support.md:3]: nit: I think we can split this into three paragraphs and make it a bit more readable

```suggestion
The OpenTelemetry Collector will be supported following a tiered platform support model to balance between the aim to support as many platforms as possible and to guarantee stability for the most important platforms. A platform is described by the pair of operating system and processor architecture family as they are defined by the Go programming language as [known operating systems and architectures for use with the GOOS and GOARCH values](https://go.dev/src/go/build/syslist.go). 

For a supported platform, the OpenTelemetry Collector is supported when the [minimum requirements](https://github.com/golang/go/wiki/MinimumRequirements) of the Go release used by the collector are met for the operating system and architecture. Each supported platform requires the naming of designated owners. The platform support for the OpenTelemetry Collector is broken into three tiers with different levels of support for each tier and aligns with the current test strategy. 

For platforms not listed as supported by any of the tiers, support cannot be assumed to be provided. While the project may accept specific changes related to these platforms, there will be no official builds, support of issues and development of enhancements or bug fixes for these platforms. Future development of project for supported platforms may break the functionality of unsupported platforms.
```

**Review by mx-psi — APPROVED** [review]
Looks good to me, just one suggestion about formatting. Thanks for your hard work on this @rrschulze!!

File: docs/platform-support.md
Comment (codeboten) [inline docs/platform-support.md:20]: The list of images may change, as it already has now that it includes s390. I worry this would get out of sync quickly. Maybe it's better to include a link to the goreleaser configuration https://github.com/open-telemetry/opentelemetry-collector-releases/blob/bf8002ec6d2109cdb4184fc6eb6f8bda59ea96a2/.goreleaser.yaml#L137?

File: docs/platform-support.md
Comment (codeboten) [inline docs/platform-support.md:24]:
```suggestion
The platform support for the OpenTelemetry Collector is broken into three tiers with different levels of support for each tier. 
```

File: docs/platform-support.md
Comment (codeboten) [inline docs/platform-support.md:38]:
```suggestion
The Tier 1 supported platforms are _guaranteed to work_. Precompiled binaries are built on the platform, fully supported for all collector add-ons (receivers, processor, exporters etc.), and continuously tested as part of the development processes to ensure any proposed change will function correctly. Build and test infrastructure is provided by the project. All tests are executed on the platform as part of automated continuous integration (CI) for each pull request and the [release cycle](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/release.md#release-schedule). Any build or test failure block the release of the collector distribution for all platforms. Defects are addressed with priority and depending on severity fixed for previous release(s) in a bug fix release.
```

File: docs/platform-support.md
Comment (codeboten) [inline docs/platform-support.md:47]:
```suggestion
Tier 2 platforms are _guaranteed to work with specified limitations_. Precompiled binaries are built and tested on the platform as part of the release cycle. Build and test infrastructure is provided by the platform maintainers. All tests are executed on the platform as far as they are applicable, and all prerequisites are fulfilled. Not executed tests and not tested collector add-ons (receivers, processors, exporters, etc.) are published on release of the collector distribution. Any build or test failure delays the release of the binaries for the respective platform but not the collector distribution for all other platforms. Defects are addressed but not with the priority as for Tier 1 and, if specific to the platform, require the support of the platform maintainers.
```

File: docs/platform-support.md
Comment (codeboten) [inline docs/platform-support.md:69]:
```suggestion
The proposed additional platform aix/ppc64 ([#19195](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19195)) will be included into Tier 3 once it's added to the OpenTelemetry Collector as platform. 
```

File: docs/platform-support.md
Comment (codeboten) [inline docs/platform-support.md:66]:
```suggestion
| linux/ppc64le |                                                                                                             |
| linux/s390x   |                                                                                                             |
```

**Review by codeboten — COMMENTED** [review]
Thanks for the writeup! Only a few comments.

File: docs/platform-support.md
Comment (rrschulze) [reply docs/platform-support.md]: Removed the reference to specific platforms for the container images and added reference to the goreleaser configuration.

**Comment by mx-psi** [comment]
> Q: what happens when a tier-3 platform build starts failing? Is the release process already resilient enough to such failures to ensure that the other platforms are still released?

@yurishkuro Right now it would make the build fail, but it would be easy enough to temporarily exclude a particular platform from the build with a small PR that we can later revert. Maybe we can make it easier, if you have concrete ideas please file an issue :)

In any case IMO this text should be read as a guide to make decisions (what to do when we have a question such as "should we delay the release because of a problem specific to platform X?") than as a prescription on what our release pipeline should look like.

**Pull request merged**
